### PR TITLE
[Inductor] support masked vectorization for the tail_loop of the 2d tiles kernel

### PR DESCRIPTION
**Pull request by jiayisunx**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #128802
* #131745
* #131155
* __->__ #130724

This PR supports masked vectorization for the tail_loop of the 2d tiles kernel to improve the performance.

Example:
```
import torch

def fn(a):
    return torch.permute(a, (2, 0, 1)).contiguous()

input = torch.randn(2, 20, 40)
compiled_fn = torch.compile(fn)

with torch.no_grad():
    for _ in range(3):
        compiled_fn(input)
```

Generated code:
- Before:
```
cpp_fused_clone_0 = async_compile.cpp_pybinding(['const float*', 'float*'], '''
#include "/tmp/torchinductor_jiayisun/z2/cz2ry4ghylembzwx7hkbanur76fi3mkiu7s6jm3zdi2amy5egq4b.h"
extern "C"  void kernel(const float* in_ptr0,
                       float* out_ptr0)
{
    {
        #pragma GCC ivdep
        for(long x0=static_cast<long>(0L); x0<static_cast<long>(32L); x0+=static_cast<long>(16L))
        {
            #pragma GCC ivdep
            for(long x1=static_cast<long>(0L); x1<static_cast<long>(32L); x1+=static_cast<long>(16L))
            {
                float tmp0[16*16] __attribute__ ((aligned (16)));
                at::vec::transpose_mxn<float,16,16>(in_ptr0 + static_cast<long>(x0 + (40L*x1)), static_cast<long>(40L), tmp0, 16);
                for (long x0_inner = 0; x0_inner < 16; x0_inner++)
                {
                    auto tmp1 = at::vec::Vectorized<float>::loadu(tmp0 + static_cast<long>(16L*x0_inner), 16);
                    tmp1.store(out_ptr0 + static_cast<long>(x1 + (40L*x0) + (40L*x0_inner)));
                }
            }
            #pragma GCC ivdep
            for(long x1=static_cast<long>(32L); x1<static_cast<long>(40L); x1+=static_cast<long>(1L))
            {
                auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + static_cast<long>(x0 + (40L*x1)), 16);
                [&]
                {
                    __at_align__ std::array<float, 16> tmpbuf;
                    tmp0.store(tmpbuf.data(), 16);
                    #pragma GCC unroll 16
                    for (long x0_inner = 0; x0_inner < 16; x0_inner++)
                    {
                        out_ptr0[static_cast<long>(x1 + (40L*x0) + (40L*x0_inner))] = tmpbuf[x0_inner];
                    }
                }
                ()
                ;
            }
        }
        #pragma GCC ivdep
        for(long x0=static_cast<long>(32L); x0<static_cast<long>(40L); x0+=static_cast<long>(1L))
        {
            #pragma GCC ivdep
            for(long x1=static_cast<long>(0L); x1<static_cast<long>(40L); x1+=static_cast<long>(1L))
            {
                auto tmp0 = in_ptr0[static_cast<long>(x0 + (40L*x1))];
                out_ptr0[static_cast<long>(x1 + (40L*x0))] = tmp0;
            }
        }
    }
}
''')


async_compile.wait(globals())
del async_compile

def call(args):
    arg0_1, = args
    args.clear()
    assert_size_stride(arg0_1, (2, 20, 40), (800, 40, 1))
    buf0 = empty_strided_cpu((40, 2, 20), (40, 20, 1), torch.float32)
    cpp_fused_clone_0(arg0_1, buf0)
    del arg0_1
    return (buf0, )
```

- After:
```
cpp_fused_clone_0 = async_compile.cpp_pybinding(['const float*', 'float*'], '''
#include "/tmp/torchinductor_jiayisun/z2/cz2ry4ghylembzwx7hkbanur76fi3mkiu7s6jm3zdi2amy5egq4b.h"
extern "C"  void kernel(const float* in_ptr0,
                       float* out_ptr0)
{
    {
        #pragma GCC ivdep
        for(long x0=static_cast<long>(0L); x0<static_cast<long>(32L); x0+=static_cast<long>(16L))
        {
            #pragma GCC ivdep
            for(long x1=static_cast<long>(0L); x1<static_cast<long>(32L); x1+=static_cast<long>(16L))
            {
                float tmp0[16*16] __attribute__ ((aligned (16)));
                at::vec::transpose_mxn<float,16,16>(in_ptr0 + static_cast<long>(x0 + (40L*x1)), static_cast<long>(40L), tmp0, 16);
                for (long x0_inner = 0; x0_inner < 16; x0_inner++)
                {
                    auto tmp1 = at::vec::Vectorized<float>::loadu(tmp0 + static_cast<long>(16L*x0_inner), 16);
                    tmp1.store(out_ptr0 + static_cast<long>(x1 + (40L*x0) + (40L*x0_inner)));
                }
            }
            #pragma GCC ivdep
            for(long x1=static_cast<long>(32L); x1<static_cast<long>(40L); x1+=static_cast<long>(8L))
            {
                float tmp0[16*8] __attribute__ ((aligned (16)));
                at::vec::transpose_mxn<float,8,16>(in_ptr0 + static_cast<long>(x0 + (40L*x1)), static_cast<long>(40L), tmp0, 8);
                for (long x0_inner = 0; x0_inner < 16; x0_inner++)
                {
                    auto tmp1 = at::vec::Vectorized<float>::loadu(tmp0 + static_cast<long>(8L*x0_inner), 8);
                    tmp1.store(out_ptr0 + static_cast<long>(x1 + (40L*x0) + (40L*x0_inner)), 8);
                }
            }
        }
        #pragma GCC ivdep
        for(long x0=static_cast<long>(32L); x0<static_cast<long>(40L); x0+=static_cast<long>(8L))
        {
            #pragma GCC ivdep
            for(long x1=static_cast<long>(0L); x1<static_cast<long>(32L); x1+=static_cast<long>(16L))
            {
                float tmp0[8*16] __attribute__ ((aligned (16)));
                at::vec::transpose_mxn<float,16,8>(in_ptr0 + static_cast<long>(x0 + (40L*x1)), static_cast<long>(40L), tmp0, 16);
                for (long x0_inner = 0; x0_inner < 8; x0_inner++)
                {
                    auto tmp1 = at::vec::Vectorized<float>::loadu(tmp0 + static_cast<long>(16L*x0_inner), 16);
                    tmp1.store(out_ptr0 + static_cast<long>(x1 + (40L*x0) + (40L*x0_inner)));
                }
            }
            #pragma GCC ivdep
            for(long x1=static_cast<long>(32L); x1<static_cast<long>(40L); x1+=static_cast<long>(8L))
            {
                float tmp0[8*8] __attribute__ ((aligned (16)));
                at::vec::transpose_mxn<float,8,8>(in_ptr0 + static_cast<long>(x0 + (40L*x1)), static_cast<long>(40L), tmp0, 8);
                for (long x0_inner = 0; x0_inner < 8; x0_inner++)
                {
                    auto tmp1 = at::vec::Vectorized<float>::loadu(tmp0 + static_cast<long>(8L*x0_inner), 8);
                    tmp1.store(out_ptr0 + static_cast<long>(x1 + (40L*x0) + (40L*x0_inner)), 8);
                }
            }
        }
    }
}
''')


async_compile.wait(globals())
del async_compile

def call(args):
    arg0_1, = args
    args.clear()
    assert_size_stride(arg0_1, (2, 20, 40), (800, 40, 1))
    buf0 = empty_strided_cpu((40, 2, 20), (40, 20, 1), torch.float32)
    cpp_fused_clone_0(arg0_1, buf0)
    del arg0_1
    return (buf0, )
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codegen/cpp.py [hunk 1]
```diff
@@ -2798,13 +2817,22 @@ def gen_transposed_tile_load_store(self, name, var, index, is_store):
...
         if is_store:
             src, dst = dst, src
             ld_src, ld_dst = ld_dst, ld_src
 
         need_define = True
-        load_or_store = f"at::vec::transpose_mxn<{DTYPE_TO_CPP[dtype]},{factor},{factor}>({src}, {ld_src}, {dst}, {ld_dst});"
+        if self.inner_is_tiling_idx ^ is_store:
+            load_or_store = (
+                f"at::vec::transpose_mxn<{DTYPE_TO_CPP[dtype]},{self.inner_num_elems},{self.outer_num_elems}>"
+                f"({src}, {ld_src}, {dst}, {ld_dst});"
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:2829]: Can you share the generated code when the tail size is used for generating this? BTW, does it work with dynamic shapes?

Comment (jiayisunx) [reply torch/_inductor/codegen/cpp.py:2817]: No it dosen't work, for dynamic shapes, the tail loop still uses scalar kernel.

Comment (jiayisunx) [reply torch/_inductor/codegen/cpp.py:2817]: I shared the generated code in the top comment.

Comment (jgong5) [reply torch/_inductor/codegen/cpp.py:2817]: > No it dosen't work, for dynamic shapes, the tail loop still uses scalar kernel.

Are you going to support dynamic shapes in future PRs?

File: torch/_inductor/codegen/cpp.py [hunk 2]
```diff
@@ -2757,11 +2757,30 @@ class CppTile2DKernel(CppVecKernel):
...
+            args,
+            num_threads,
+            tiling_factor,
+            tiling_indices[1],
+            tiling_dtype,
+            inner_tail_size,
         )
         self.tiling_indices = tiling_indices
+        self.inner_tail_size = inner_tail_size
+        self.outer_tail_size = outer_tail_size
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:2780]: Can we add assertion checks here to make sure `inner_tail_size` and `outer_tail_size` not dynamic?

File: torch/_inductor/codegen/cpp.py [hunk 3]
```diff
@@ -3602,25 +3647,72 @@ def select_tiling(dtype: torch.dtype = torch.float):
...
                     inner_tail_loop,
                 ) = outer_main_loop.split_with_tiling(
                     tiling_indices[1] - tiling_indices[0], factor=tiling_factors[0]
                 )
+                tile2d_kernel = codegen_kernel(
+                    CppTile2DKernel, tiling_factors[0], tiling_indices, vec_dtype
+                )
                 inner_main_loop.set_kernel(tile2d_kernel)
-                inner_tail_loop.set_kernel(vec_kernel)
+                if could_masked_vec:
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:3664]: Can we combine the logic under two `if could_masked_vec` into one?

File: torch/_inductor/codegen/cpp.py [hunk 4]
```diff
@@ -3602,25 +3647,72 @@ def select_tiling(dtype: torch.dtype = torch.float):
...
+                    )
+                    masked_tile2d_kernel3 = codegen_kernel(
+                        CppTile2DKernel,
+                        tiling_factors[0],
+                        tiling_indices,
+                        vec_dtype,
+                        inner_tail_loop_of_outer_tail_loop.steps,
+                        outer_tail_loop.steps,
+                    )
+                    inner_tail_loop_of_outer_tail_loop.set_kernel(masked_tile2d_kernel3)
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:3713]: Can we simplify the code with loops or functions to avoid dup?

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: aten/src/ATen/cpu/vec/vec512/vec512_float.h [hunk 5]
```diff
@@ -793,6 +793,80 @@ inline void transpose_mxn<float, 16, 16>(
...
+    input[i] = _mm512_maskz_loadu_ps(src_mask, &src[i * ld_src]);
+  }
+  for (; i < 16; ++i) {
+    // Not really needed but to avoid uninitialized variable warning.
+    // Shouldn't be much overhead because xor can be executed in parallel with
+    // other instructions.
+    input[i] = _mm512_setzero_ps();
+  }
+
+  // unpacking and interleaving 32-bit elements
```
Comment (jgong5) [inline aten/src/ATen/cpu/vec/vec512/vec512_float.h:817]: The logic below is same as that in `transpose_mxn<float, 16, 16>`. Can we factor out a common function to share and avoid dup?

File: aten/src/ATen/cpu/vec/vec512/vec512_float.h
Comment (jgong5) [inline aten/src/ATen/cpu/vec/vec512/vec512_float.h:1]: Are you going to support avx2 as well?

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_inductor/codegen/cpp.py (hunk 1 above)
Comment (jiayisunx) [reply torch/_inductor/codegen/cpp.py:2817]: We will consider supporting dynamic shapes in the future.

File: torch/_inductor/codegen/cpp.py (hunk 2 above)
Comment (jiayisunx) [reply torch/_inductor/codegen/cpp.py:2780]: I checked `has_free_symbols `in `CppVecKernelChecker`. for dynamic shapes, `could_masked_vec ` is False and the tail loop still uses scalar kernel.

File: torch/_inductor/codegen/cpp.py (hunk 3 above)
Comment (jiayisunx) [reply torch/_inductor/codegen/cpp.py:3665]: Done, thanks!

File: torch/_inductor/codegen/cpp.py (hunk 4 above)
Comment (jiayisunx) [reply torch/_inductor/codegen/cpp.py:3710]: There are three types of tail loops for the 2D tiles kernel: `inner_tail_loop`(inner_tail_loop of outer main loop) , `inner_main_loop_of_outer_tail_loop`, `inner_tail_loop_of_outer_tail_loop`. These three types of tail loops require `codegen_kernel `and `set_kernel` respectively.

File: aten/src/ATen/cpu/vec/vec512/vec512_float.h (hunk 5 above)
Comment (jiayisunx) [reply aten/src/ATen/cpu/vec/vec512/vec512_float.h:602]: I removed this template specialization function(`transpose_mxn<float, 16, 16>`).

File: aten/src/ATen/cpu/vec/vec512/vec512_float.h
Comment (jiayisunx) [reply aten/src/ATen/cpu/vec/vec512/vec512_float.h:1]: Yes

File: torch/_inductor/codegen/cpp.py (hunk 4 above)
Comment (jgong5) [reply torch/_inductor/codegen/cpp.py:3710]: I mean you can perhaps further simplify the code. This code pattern now looks like an unrolled loop...

Comment (jiayisunx) [reply torch/_inductor/codegen/cpp.py]: Thanks for your advice, I have simplified the code.

**Review by jgong5 — APPROVED** [review]

File: aten/src/ATen/cpu/vec/vec512/vec512_bfloat16.h [hunk 6]
```diff
@@ -1401,6 +1425,30 @@ inline void transpose_mxn<Half, 32, 32>(
   }
 }
 
+template <typename T, int M, int N,
+          typename std::enable_if_t<std::is_same<T, Half>::value && ((M < 32 && M != 16) || (N < 32 && N != 16)), int> = 0>
+inline void transpose_mxn(const Half* src, int64_t ld_src, Half* dst, int64_t ld_dst) {
+  // load from src
+  __mmask32 src_mask = (1 << N) - 1;
```
Comment (leslie-fang-intel) [inline aten/src/ATen/cpu/vec/vec512/vec512_bfloat16.h:1432]: Looks like same implementation between `Half` and `BFloat16`. Can we move it into a common util function?

File: torch/_inductor/codegen/cpp.py [hunk 7]
```diff
@@ -2757,11 +2757,30 @@ class CppTile2DKernel(CppVecKernel):
...
+            tiling_indices[1],
+            tiling_dtype,
+            inner_tail_size,
         )
         self.tiling_indices = tiling_indices
+        self.inner_tail_size = inner_tail_size
+        self.outer_tail_size = outer_tail_size
+        self.inner_num_elems = inner_tail_size if inner_tail_size else tiling_factor
+        self.outer_num_elems = outer_tail_size if outer_tail_size else tiling_factor
+        self.inner_is_tiling_idx = True
```
Comment (leslie-fang-intel) [inline torch/_inductor/codegen/cpp.py:2783]: The init value should be `None`?

**Review by leslie-fang-intel — APPROVED** [review]

File: aten/src/ATen/cpu/vec/vec512/vec512_bfloat16.h (hunk 6 above)
Comment (jiayisunx) [reply aten/src/ATen/cpu/vec/vec512/vec512_bfloat16.h:1432]: We must specialize this function here to ensure the priority of this function.

File: torch/_inductor/codegen/cpp.py (hunk 7 above)
Comment (jiayisunx) [reply torch/_inductor/codegen/cpp.py:2783]: In the init function `tiling_idx` is `tiling_indices[1]`, and `self.inner_is_tiling_idx = True`.

File: aten/src/ATen/cpu/vec/vec512/vec512_bfloat16.h (hunk 6 above)
Comment (leslie-fang-intel) [reply aten/src/ATen/cpu/vec/vec512/vec512_bfloat16.h:1432]: We can still provide a util function as
`
template <typename T, typename std::enable_if_t<std::is_same_v<T, Half> || std::is_same_v<T, BF16>>>
transpose_mxn_32_32(const T* src, int64_t ld_src, T* dst, int64_t ld_dst)
`
then reuse this util function for BF16 and FP16, it can save lines of code.

File: torch/_inductor/codegen/cpp.py (hunk 7 above)
Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp.py:2783]: oh, yes, we use `tiling_indices[1]` to init `tiling_idx` in the tile2D kernel.

**Comment by jiayisunx** [comment]
@jansel , could you please review this PR? Thanks!

**Review by jansel — APPROVED** [review]
